{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# imported libraries\n",
    "# ********************************************************************************************\n",
    "import csv\n",
    "import pandas as pd\n",
    "from scipy.stats import chi2_contingency\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from statistics import median, mean\n",
    "\n",
    "# ********************************************************************************************\n",
    "# Mapping Human IDs to Population Codes\n",
    "# ********************************************************************************************\n",
    "\n",
    "# Main Path Function\n",
    "    #  Input: a file of raw, unparsed human profile data with correspondign human IDs\n",
    "    #  Output: a dictionary where the key is the human ID and the value is the corresponding population code\n",
    "def get_humanID_to_PopulationCode(file_in):\n",
    "\n",
    "    # Open the CSV file with the open() function\n",
    "    # This returns a file object that you can use to read the file\n",
    "    with open(file_in) as csvfile:\n",
    "        \n",
    "        # Use the csv.reader() function to read the file\n",
    "        # This returns an iterable object that you can loop over\n",
    "        data = csv.reader(csvfile)\n",
    "\n",
    "        rawRowDataAsList = []\n",
    "\n",
    "        # Loop over the rows in the file\n",
    "        for row in data:\n",
    "            rawRowDataAsList.append(row)\n",
    "\n",
    "        humanIDInfoHeader = rawRowDataAsList[0][0].split(\"\\t\")\n",
    "        del rawRowDataAsList[0]\n",
    "\n",
    "        humanInfoParsed = []\n",
    "\n",
    "        for rawHumanInfo in rawRowDataAsList:\n",
    "            singleHumanInfoParsed = rawHumanInfo[0].split(\"\\t\")\n",
    "            humanInfoParsed.append(singleHumanInfoParsed)\n",
    "\n",
    "        humanID_to_PopulationCode = {}\n",
    "\n",
    "        for humanInfo in humanInfoParsed:\n",
    "            humanID = humanInfo[0]\n",
    "            populationCode = humanInfo[3]\n",
    "            humanID_to_PopulationCode[humanID] = populationCode\n",
    "\n",
    "    return humanID_to_PopulationCode\n",
    "\n",
    "# ********************************************************************************************\n",
    "# Merge genotypes with population codes by corresponding human IDs\n",
    "# ********************************************************************************************\n",
    "# Helper Function\n",
    "    # Input: a dictionary of population codes to genotypes that have not be summed yet for a single snp\n",
    "    # Output: a dictionary of the summed genotypes for the corresponding population code for a single snp --> population code : [recessive, heterozygous, dominant]\n",
    "def get_populationCode_to_geneFrequencies_for_single_snp(snp):\n",
    "    populationCode_to_geneFrequencies = {}\n",
    "\n",
    "    for populationCode_genotype in snp:\n",
    "\n",
    "        populationCode = populationCode_genotype[0]\n",
    "        genotype = populationCode_genotype[1]\n",
    "\n",
    "        if genotype == 0:\n",
    "                genotypeList = [1, 0, 0]\n",
    "        elif genotype == 1:\n",
    "            genotypeList = [0, 1, 0]\n",
    "        elif genotype == 2:\n",
    "            genotypeList = [0, 0, 1]\n",
    "\n",
    "        if populationCode not in populationCode_to_geneFrequencies.keys():\n",
    "            populationCode_to_geneFrequencies[populationCode] = genotypeList\n",
    "\n",
    "        elif populationCode in populationCode_to_geneFrequencies.keys():\n",
    "            currentGenotypeList = populationCode_to_geneFrequencies.get(populationCode)\n",
    "\n",
    "            # Use a list comprehension to add the elements from the two lists together\n",
    "            totalGenotypeList = [x + y for x, y in zip(currentGenotypeList, genotypeList)]\n",
    "\n",
    "            populationCode_to_geneFrequencies[populationCode] = totalGenotypeList\n",
    "\n",
    "    return populationCode_to_geneFrequencies\n",
    "\n",
    "# Main Path Function\n",
    "    # Input: \n",
    "        # humanIDs_control --> list of human IDs\n",
    "        # SNPData_control --> list of SNP data\n",
    "        # humanID_to_PopulationCode_sample --> a dictionary mapping human IDs to population codes\n",
    "    # Output: a dataframe where the column hears are the population code of the correspondign human ID and the row headers are the SNP\n",
    "def create_df_of_genotypes_populationCodes(humanIDs_control, SNPData_control, humanID_to_PopulationCode_sample):\n",
    "    populationCodes = []\n",
    "    for humanID in humanIDs_control:\n",
    "        populationCode = humanID_to_PopulationCode_sample.get(humanID)\n",
    "        populationCodes.append(populationCode)\n",
    "\n",
    "    # Create the dataframe with specified column names\n",
    "    df = pd.DataFrame(SNPData_control, columns=populationCodes)\n",
    "\n",
    "    return df\n",
    "\n",
    "# Main Path Function \n",
    "    # Input: a dataframe of snps with columns = population codes, rows = snps, values = genotypes\n",
    "    # Output: a list of lists --> where each outer list is a snp and the inner list is [population code, genotype] that are NOT summed\n",
    "def get_snps_populationCode_to_genotype_control(df):\n",
    "    snps_populationCode_to_genotype_control = []\n",
    "\n",
    "    # Loop through every row in the dataframe\n",
    "    # i = 0\n",
    "    for index, row in df.iterrows():\n",
    "        # Loop through every column name in the dataframe\n",
    "        # j = 0\n",
    "        snp_populationCode_to_genotype = []\n",
    "        for x in range(len(df.columns)):\n",
    "            # Print the values in each column\n",
    "            # print(row[x])\n",
    "        # for col in df.columns:\n",
    "            # Print the values in each column\n",
    "            # print(type())\n",
    "            # populationCode_to_genotype[df.columns[x]] = row[x]\n",
    "            snp_populationCode_to_genotype.append([df.columns[x], row[x]])\n",
    "        #     print(\"--\", i, j, \"Population Code:\", df.columns[x], \"Genotype:\", row[x], \"...END...\")\n",
    "        #     j += 1\n",
    "        # i += 1\n",
    "        snps_populationCode_to_genotype_control.append(snp_populationCode_to_genotype)\n",
    "\n",
    "    # print(\"populationCode_to_genotype_control\\n\", snps_populationCode_to_genotype_control)\n",
    "    return snps_populationCode_to_genotype_control\n",
    "\n",
    "# Main Path function\n",
    "    # Input: a list of lists --> where each outer list is a snp and the inner list is [population code, genotype] that are NOT summed\n",
    "    # Output: a list of dictionaries where each index in the list is a snp and each dictionary's key is a population code and corresponding value is a list of the summed [recessive, heterozygous, dominant]\n",
    "def get_snps_list_populationCode_to_geneFrequencies_control(snps_populationCode_to_genotype_control):\n",
    "    snps_list_populationCode_to_geneFrequencies_control = []\n",
    "\n",
    "    for snp in snps_populationCode_to_genotype_control:\n",
    "\n",
    "        populationCode_to_geneFrequencies = get_populationCode_to_geneFrequencies_for_single_snp(snp)\n",
    "        snps_list_populationCode_to_geneFrequencies_control.append(populationCode_to_geneFrequencies)\n",
    "\n",
    "    return snps_list_populationCode_to_geneFrequencies_control\n",
    "\n",
    "# ********************************************************************************************\n",
    "# Perform Chi-square Statistical test \n",
    "# ********************************************************************************************\n",
    "\n",
    "# Helper Function\n",
    "    # Input: a dictionary of population codes to summed genetype frequencies\n",
    "    # Output: a dataframe where the columen headers are the genotypes and the row headers are the population codes\n",
    "def create_df_from_populationCode_to_geneFrequencies(populationCode_to_geneFrequencies):\n",
    "    \n",
    "    geneFrequenciesList = []\n",
    "    populationCodesList = []\n",
    "    for key, value in populationCode_to_geneFrequencies.items():\n",
    "        geneFrequenciesList.append(value)\n",
    "        populationCodesList.append(key)\n",
    "\n",
    "    # Create a list of column names\n",
    "    columns = ['Recesssive', 'Heterozygous', 'Dominant']\n",
    "\n",
    "    # Convert the list of lists to a DataFrame\n",
    "    df = pd.DataFrame(geneFrequenciesList, columns=columns, index=populationCodesList)\n",
    "\n",
    "    # Print the resulting DataFrame\n",
    "    # print(df)\n",
    "\n",
    "    return df\n",
    "\n",
    "# Helper Function\n",
    "    # Input: a dataframe of summed genotypes and population codes\n",
    "    # Output: \n",
    "        # [chi2, p] --> a list of the chi square stat and the p value\n",
    "        #  -1 --> no variance in snp\n",
    "def perform_chi_square_test(df):\n",
    "    # Perform the chi-square test\n",
    "    try: \n",
    "        chi2, p, dof, expected = chi2_contingency(df)\n",
    "        return [chi2, p]\n",
    "    except:\n",
    "        return -1\n",
    "\n",
    "# Main Path Function\n",
    "    # Input: a list of dictionaries where each index in the list is a snp and each dictionary's key is a population code and corresponding value is a list of the summed [recessive, heterozygous, dominant]\n",
    "    # Output: a list of chi2 values or -1 if chi2 could not be calculated\n",
    "def get_chi2_p_forEachSNP_list(snps_list_populationCode_to_geneFrequencies_control):\n",
    "\n",
    "    chi2_p_list = []\n",
    "\n",
    "    for snp in snps_list_populationCode_to_geneFrequencies_control:\n",
    "\n",
    "        df = create_df_from_populationCode_to_geneFrequencies(snp)\n",
    "        # print(df)\n",
    "        chi2_p = perform_chi_square_test(df)\n",
    "        chi2_p_list.append(chi2_p)\n",
    "    \n",
    "    return chi2_p_list\n",
    "\n",
    "\n",
    "# ********************************************************************************************\n",
    "# Additional Analysis from Chi-square Statistical test \n",
    "# ********************************************************************************************\n",
    "    \n",
    "# Main Path Function\n",
    "    # Input: raw chi2 and p-value list\n",
    "    # Output:\n",
    "        # chi2_list: list of chi squared scores\n",
    "        # no_chi2_percentage: a percenatge of the numbe rof uncomputable chi square snps compared to the total snps tested\n",
    "        # p_value_list: list of p values\n",
    "def remove_uncomputable_chi2_values_and_calculate_percentage(snps_chi2_p_list):\n",
    "    totalSNPS = len(snps_chi2_p_list)\n",
    "\n",
    "    no_chi2_cnt = 0\n",
    "\n",
    "    chi2_list = []\n",
    "    p_value_list = []\n",
    "\n",
    "    for chi2_p in snps_chi2_p_list:\n",
    "\n",
    "        if chi2_p == -1:\n",
    "            no_chi2_cnt += 1\n",
    "\n",
    "        else:\n",
    "            chi2 = chi2_p[0]\n",
    "            p_value = chi2_p[1]\n",
    "\n",
    "            chi2_list.append(chi2)\n",
    "            p_value_list.append(p_value)\n",
    "\n",
    "\n",
    "    no_chi2_percentage = (no_chi2_cnt/totalSNPS)*100\n",
    "\n",
    "    # print(\"no_chi2_percentage\\n\", no_chi2_percentage)\n",
    "\n",
    "    # print(len(chi2_list))\n",
    "    # print(\"chi2_list\\n\", chi2_list)\n",
    "    # print(\"p_value_list\\n\", p_value_list)\n",
    "\n",
    "    return chi2_list, no_chi2_percentage, p_value_list\n",
    "\n",
    "# Helper Function\n",
    "    # Input: list chi sqaured scores \n",
    "    # Output: the value of the 95th percentile of the chi2 scores for the snps\n",
    "def get_95th_percentile_chi2_score(chi2_list):\n",
    "    \n",
    "    p99 = np.percentile(chi2_list, 99)\n",
    "    p95 = np.percentile(chi2_list, 95)\n",
    "    med = median(chi2_list)\n",
    "    avg = mean(chi2_list)\n",
    "    # Print the value at the 95th percentile\n",
    "    # print(p95)\n",
    "    return p99, p95, med, avg\n",
    "\n",
    "# Main Path Function\n",
    "    # Input: list chi sqaured scores \n",
    "    # Output: None\n",
    "def create_hist(chi2_list):\n",
    "    # Create the histogram with 5 bins\n",
    "    plt.hist(chi2_list, bins=40)\n",
    "    # add a title\n",
    "    plt.title(\"Chi Square Distribution for SNPs\")\n",
    "    # add x and y axis labels\n",
    "    plt.xlabel(\"Chi Square Value\")\n",
    "    plt.ylabel(\"Chi Square Count per bin range\")\n",
    "    # Show the histogram\n",
    "    plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Single Cancer SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ********************************************************************************************\n",
    "# Gene data parsing and organization\n",
    "# ********************************************************************************************\n",
    "\n",
    "# Helper Function\n",
    "    # Input: a list that contains raw SNPs for a number of HumanIDs\n",
    "    # Output: a list that contains SNPs for number of HumanIDs that have been converted based on genotype \n",
    "        # 0 = homozygous recessive\n",
    "        # 1 = heterozygous\n",
    "        # 2 = homozygous dominant\n",
    "def convert_rawSNPData_to_Standard(snp_list):\n",
    "\n",
    "    standard_snp_list = []\n",
    "\n",
    "    for humanGene in snp_list:\n",
    "        if humanGene == \"0|0\":\n",
    "            standard_snp_list.append(0)\n",
    "        elif humanGene == \"1|0\" or humanGene == \"0|1\":\n",
    "            standard_snp_list.append(1)\n",
    "        elif humanGene == \"1|1\":\n",
    "            standard_snp_list.append(2)\n",
    "\n",
    "    return standard_snp_list\n",
    "\n",
    "# Main Path Function\n",
    "    # Input: a file containing the raw gene data for numerous SNPs\n",
    "    # Output: \n",
    "        # humanIDs -> a list of human IDs for the SNPS\n",
    "        # geneData -> a list of raw, unparsed SNP gene data for the human IDs\n",
    "def get_humanIDsList_SNPsList_control(file_in):\n",
    "\n",
    "    # Open the CSV file with the open() function\n",
    "    # This returns a file object that you can use to read the file\n",
    "    with open(file_in) as csvfile:\n",
    "\n",
    "        # Use the csv.reader() function to read the file\n",
    "        # This returns an iterable object that you can loop over\n",
    "        data = csv.reader(csvfile)\n",
    "\n",
    "        rawRowDataAsList = []\n",
    "\n",
    "        # Loop over the rows in the file\n",
    "        for row in data:\n",
    "            rawRowDataAsList.append(row)\n",
    "\n",
    "        # HUMAN IDs \n",
    "        rawHumanIDs = rawRowDataAsList[19][0].split(\"\\t\")\n",
    "        # print('rawHumanIDs\\n', rawHumanIDs)\n",
    "        humanIDs = rawHumanIDs[9: len(rawHumanIDs)-1]\n",
    "        # print('humanIDs\\n', humanIDs)\n",
    "\n",
    "        # GENE DATA\n",
    "        rawSNPData = rawRowDataAsList[20]\n",
    "        # print('rawSNPData\\n', rawSNPData)\n",
    "        rawGeneRowParsed = rawSNPData[0].split(\"\\t\")\n",
    "        snp_list_raw = rawGeneRowParsed[9:len(rawGeneRowParsed)-1]\n",
    "        single_snp_list = convert_rawSNPData_to_Standard(snp_list_raw)\n",
    "\n",
    "    return (humanIDs, [single_snp_list])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# *************************************************************************************\n",
    "# Analysis for single target cancer SNP\n",
    "# **************************************************************************************\n",
    "\n",
    "def main_single_snp(file_in, humanID_to_PopulationCode):\n",
    "\n",
    "    # 1. Gene data parsing and organization\n",
    "    humanIDs, SNPsDataList = get_humanIDsList_SNPsList_control(file_in)\n",
    "\n",
    "    # print(humanIDs)\n",
    "    # print(SNPsDataList)\n",
    "\n",
    "    # 2. Merge genotypes with population codes by corresponding human IDs\n",
    "    df_populationCodes_genotypes = create_df_of_genotypes_populationCodes(humanIDs, SNPsDataList, humanID_to_PopulationCode)\n",
    "\n",
    "    snps_populationCode_to_genotype = get_snps_populationCode_to_genotype_control(df_populationCodes_genotypes)\n",
    "\n",
    "    snps_list_populationCode_to_geneFrequencies = get_snps_list_populationCode_to_geneFrequencies_control(snps_populationCode_to_genotype)\n",
    "\n",
    "    # 3. Perform Chi-square Statistical test \n",
    "    chi2_p_list_raw = get_chi2_p_forEachSNP_list(snps_list_populationCode_to_geneFrequencies)\n",
    "\n",
    "    print(chi2_p_list_raw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "humanID_to_PopulationCode = get_humanID_to_PopulationCode('rawData/igsr_samples.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1016.5887020893715, 2.6643478095338636e-180]]\n"
     ]
    }
   ],
   "source": [
    "main_single_snp(\"rawData/rs4713266.csv\", humanID_to_PopulationCode)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13 (main, Aug 25 2022, 18:29:29) \n[Clang 12.0.0 ]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "e862235458dcc43ce60f6439dfba4578949053168f0a6b8272bc7c7b0264254d"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
